### ISSUE TITLE
epic(refactor): stage-axis audit findings — TwoPhaseBase + PuLIDMixin + typed errors (per lyra #1277)

### ISSUE THREAD

**Issue by MickaelV0** [Roxabi/imageCLI#89]
**Source audit:** [`artifacts/analyses/stage-axis-audit-2026-05-20.md`](https://github.com/Roxabi/imageCLI/blob/main/artifacts/analyses/stage-axis-audit-2026-05-20.md)
**Framework:** lyra [#1277](https://github.com/Roxabi/lyra/issues/1277) — stage-axis decomposition
**Reference doc:** `~/projects/lyra/artifacts/analyses/1277-stage-axis-refactor-strategy.mdx`

---

## TL;DR — Verdict

imageCLI is **definitively on the target axis** with high N (9 engines, growing) and high M (7+ concerns per 2-phase engine file). The cascade is **latent, not yet active**: recent refactors (engine subpackage split, `helpers.py`, `_pulid/` sub-package, `nvfp4/` sub-package, `validators.py` extraction) show the team fighting the pattern one file at a time without naming the root cause.

The 3-engine quantization family (`flux2-klein`, `flux2-klein-fp8`, `flux2-klein-fp4`) is the hottest N×M zone: **75% file similarity** between flux2-klein and fp8; **76 lines word-for-word identical** across all 3 quant files. The next quant variant (e.g., quanto int8 for Ampere, or NVFP4 + LoRA) would create a 4th copy of the same block.

imageCLI has the **strongest cascade pressure** of the 3-CLI fleet (llmCLI / voiceCLI / imageCLI). This issue tracks 5 findings + 4 cross-repo follow-ups.

---

### Finding 1 — 2-Phase infrastructure is N×M: 5 methods × 3 files (LATENT)

**Evidence:** `flux2_klein.py:121-277`, `flux2_klein_fp8.py:117-270`, `flux2_klein_fp4.py:119-247` — 5 methods each, **76 lines word-for-word identical**.

**Similarity ratios (SequenceMatcher):**
- `flux2_klein.py` vs `flux2_klein_fp8.py`: **75.0%** identical
- `flux2_klein_fp4.py` vs `flux2_klein_fp8.py`: **56.9%** identical
- `flux2_klein.py` vs `flux2_klein_fp4.py`: **52.7%** identical

**Duplicated methods:** `encode_and_generate`, `generate_from_embeddings`, `encode_prompt`, `load_for_encode`, `start_generation_phase` — all 5 duplicated × 3 files. The `start_generation_phase` gc teardown (`text_encoder.to("cpu") + empty_cache + gc.collect`) is **verbatim** in all 3 (`:204-206`, `:211-213`, `:196-198`).

**Risk:** Latent. Next quant variant adds 4th copy. A bug in `encode_prompt` return format or `generate_from_embeddings` kwargs propagates to 3 files.

**Fix:** Extract `TwoPhaseBase` concrete base with:
- `encode_and_generate()` implemented once using `self._pipe` + `self._build_pipe_kwargs()`
- `generate_from_embeddings()` implemented once
- `_teardown_encoder_phase()` helper for the gc teardown
- Engines provide only `_load_pipeline()` and `_set_execution_device()`

**~150 LOC reduction** (3 files × 50 LOC saved).

### Finding 2 — `_set_execution_device` / `_execution_device_override`: 4 impls × 3 strategies (ACTIVE)

- `flux2_klein.py:108-116` (inline, in `load_all_on_gpu`) + `:215-222` (inline, in `start_generation_phase`) — **2 inline copies in one file**
- `flux2_klein_fp8.py:101-115` — extracted `_set_execution_device()` with assert
- `flux2_klein_fp4.py:95-107` — extracted `_set_execution_device()` without assert
- `pulid_flux2_klein_fp4.py:79-94` — different strategy: **anonymous subclass to isolate the patch**

The anonymous-subclass approach is strictly safer (scoped to instance, not class), but was not backported. **Active cascade risk** — a bug in the monkey-patch logic requires touching 4 sites.

**Fix:** Adopt the anonymous-subclass strategy once in `TwoPhaseBase` or `helpers._set_execution_device(pipe)`. Remove all per-engine copies.

### Finding 3 — `_check_requirements` duplicated between fp4 engines (LATENT)

`flux2_klein_fp4.py:40-57` and `pulid_flux2_klein_fp4.py:54-76` are near-identical (14-line diff, only error message strings differ). A new fp4 engine variant would copy this block again.

**Fix:** Extract `_check_nvfp4_requirements()` to `engines/nvfp4/__init__.py` (sub-package already exists with `quantize.py`, `state_dict.py`). Call from both engines.

### Finding 4 — Socket daemon leaks `str(exc)` to clients on 3 bus-bound paths (ACTIVE) — cross-repo pattern, see Cross-repo #3

`daemon_handlers.py:83,154,238` — identical `{"ok": False, "error": str(exc)}` pattern in 3 separate exception handlers. An OOM, path access error, or GGUF download failure is sent verbatim to the socket client.

The NATS path correctly sanitizes via `_map_exception_to_error` → typed `WorkerError`. Two-tier protocol with divergent error contracts.

**Fix:** See "Cross-repo follow-ups" #3 below (shared sanitization helper).

### Finding 5 — String-matching on `str(exc)` for exception type dispatch (ACTIVE)

`nats/validators.py:205`:
```python
if isinstance(exc, ValueError) and "Unknown engine" in str(exc):
    return "unknown_engine", str(exc).split(": ", 1)[-1]
```

String-matching on exception message to detect type. If `registry.py:50` message changes ("Unknown engine" → "Engine not found"), the `unknown_engine` code path silently falls through to `generation_failed`, making all unknown-engine errors retryable.

**Fix:** Introduce `UnknownEngineError(ValueError)` in `engine/registry.py`. Check `isinstance(exc, UnknownEngineError)` in `_map_exception_to_error`.

---

### Cross-repo #1 — Promote NATS subjects to `roxabi-contracts` constants

**Evidence in imageCLI:** `nats/adapter.py:23-24`:
```python
SUBJECT = "lyra.image.generate.request"
HEARTBEAT_SUBJECT = "lyra.image.heartbeat"
```

Subjects hardcoded. llmCLI is the reference — imports `roxabi_contracts.llm.SUBJECTS`.

**Steps:**
1. **Upstream (lyra):** add `voice.SUBJECTS` + `image.SUBJECTS` modules in `packages/roxabi-contracts/`
2. **imageCLI:** replace `nats/adapter.py:23-24` literals with imports

Blocked-by: lyra issue (TBD, will be created as parent).

### Cross-repo #2 — Align `roxabi-nats` to `branch=staging` (was: pinned `v0.4.1`)

**Current state:** `pyproject.toml`:
```toml
roxabi-nats = { ..., tag = "roxabi-nats/v0.4.1" }
```

llmCLI and voiceCLI use `branch = "staging"`. imageCLI is the outlier.

**Why align:** automatic propagation of upstream contract changes (subjects, schema bumps, worker options like `wait_ready=False`). Removes the cross-repo sibling-fix pattern observed in commits `c241600` (imageCLI) + `b61a998` (voiceCLI) where the same `wait_ready=False` opt-in was implemented separately in each repo.

**Trade-off:** loses version-pin safety (force-pushed tag risk). Mitigated by hash-locked `uv.lock` + the controlled lyra monorepo.

**Fix:** 1-line change in `pyproject.toml` — `tag = "roxabi-nats/v0.4.1"` → `branch = "staging"`. Run `uv sync` and verify all tests pass.

### Cross-repo #3 — Declare `roxabi-contracts` as direct dependency

**Subsumes:** #86 (folded into this issue).

**Evidence:** imageCLI imports `roxabi_contracts` directly at `nats/adapter.py:14-15` (`WorkerError`, `ImageResponse`) but does **not** declare it in `pyproject.toml` — transitive via `roxabi-nats` only.

**Risk:** If `roxabi-nats` stops re-exporting `roxabi-contracts` transitively, imageCLI's NATS layer breaks silently. (#86 noted the same concern but contained a factual error — claimed voiceCLI declared it explicitly, which is not the case. Both voiceCLI and imageCLI need this fix.)

**Fix (local, 1 line):** Add `"roxabi-contracts"` under `[project.optional-dependencies] nats` + add corresponding `[tool.uv.sources]` entry pointing to the same `branch = "staging"` (after Cross-repo #2 alignment). Mirror of llmCLI pattern.

→ This issue will close #86 as subsumed.

### Cross-repo #4 — `str(exc)` socket leak (shared anti-pattern × 3 daemons)

**Evidence in imageCLI:** `daemon_handlers.py:83,154,238` (Finding 4). Same anti-pattern exists in llmCLI (2 sites) and voiceCLI (3 sites). NATS path in all 3 projects is correctly typed via `WorkerError` — only the socket path leaks.

**Shared fix (upstream + local):**
1. **Upstream (roxabi-nats):** add `sanitize_for_wire(exc) -> str` helper (truncate, redact stack, drop file paths)
2. **imageCLI:** replace 3 `str(exc)` sites with `sanitize_for_wire(exc)` in `daemon_handlers.py`
3. **Pre-push lint rule:** flag `str(exc)` in any file containing `send_json`/`nats.publish`

Blocked-by: lyra issue (TBD, will be created as parent).

---

## DEBT slugs (per lyra #1277 inventory)

| Slug | Present? | Evidence |
|---|---|---|
| `adapter-magic-constants` | **present** | NATS subjects hardcoded (Cross-repo #1), `"Unknown engine"` string match (Finding 5) |
| `boundary-broad-catch` | partial | 12 `except Exception` sites total, 4 in `daemon_handlers.py` convert exc → `str(exc)` for socket (Finding 4) |
| `complexity-residual` | **present** | `nats/adapter.py:357 LOC` (over 300-line gate) |
| `wiring-bootstrap-deps` | no | — |
| Others | no | — |

**`artifacts/debt/` does not exist** — no formal debt tracking. Inline tags: **1 TODO** (`engine/registry.py:61` — `TODO(#72): remove legacy singular-kwargs branch after one release cycle`).

---

## Top 3 actions (prioritized)

1. **Extract `TwoPhaseBase` with concrete `encode_and_generate` + `generate_from_embeddings` + `_teardown_encoder_phase`** (Finding 1) — removes ~150 LOC, prevents 4th quant variant copy. Targets the hottest N×M zone.

2. **Fix `str(exc)` socket leaks + introduce `UnknownEngineError`** (Findings 4 + 5) — removes 3 active bus leaks + fixes fragile string-matching dispatch. Internal, no external contract change.

3. **Extract `_check_nvfp4_requirements()` to `engines/nvfp4/__init__.py` + `PuLIDMixin`** (Finding 3 + bonus) — removes 2 duplicated check blocks + InsightFace/EVA-CLIP loading blocks (3 PuLID engines share verbatim 7-line blocks at `pulid_flux2_klein.py:84-91`, `pulid_flux2_klein_fp4.py:151-158`, `pulid_flux1_dev.py:89-96`).

---

## Bonus — Special angle: PuLID composition path

**Current state:** `pulid_flux2_klein.py` overrides `generate()` to inject PuLID tokens and wraps `super().generate()` in a patch/unpatch sandwich. `pulid_flux2_klein_fp4.py` reimplements the same sandwich plus manual prompt encoding. InsightFace + EVA-CLIP loading is **duplicated 3×**.

**Clean composition path — `PuLIDMixin`:**
1. Holds `_pulid`, `_insightface`, `_eva_clip` state
2. Provides `_load_face_models(insightface_dir, pulid_path, eva_clip_model)` once
3. Provides `_extract_id_tokens(refs)` — delegates to `_pulid.extract_id_tokens` for Flux2 or the custom 63-line impl for Flux1 (architecturally different — not trivially unified, but InsightFace block is)
4. Provides `_generate_with_pulid(prompt, refs, pulid_strength, **kwargs)` with the patch/unpatch pattern

`pulid_flux2_klein.py` and `pulid_flux2_klein_fp4.py` compose: `PuLIDMixin + Flux2KleinEngine` vs `PuLIDMixin + Flux2KleinFP4Engine`. **~80 LOC reduction across 3 PuLID files.**

---

## Out of scope

- Stage-axis pivot proper (the lyra-style 7-phase rewrite) — too heavy for current scale; the concrete `TwoPhaseBase` + `PuLIDMixin` extractions deliver 80% of the value
- Replacing `ImageEngine` ABC with stage-pipeline ABC — current ABC is appropriate
- `engine/registry.py:61` TODO (remove legacy singular-kwargs branch) — tracked separately as #72

---

## Background

Audit conducted 2026-05-20 as part of cross-repo lyra #1277 follow-up. Parallel audits filed:

- llmCLI: [#53](https://github.com/Roxabi/llmCLI/issues/53)
- voiceCLI: [#162](https://github.com/Roxabi/voiceCLI/issues/162)
- lyra parent (cross-repo): (TBD, will own Cross-repo #1 + #4)

**Closes:** #86 (subsumed — folded into Cross-repo #3 above).

**Comment by MickaelV0** [comment]
**Parent (cross-repo):** Roxabi/lyra#1291 — owns the upstream additions:
- `image.SUBJECTS` module in roxabi-contracts (unblocks Cross-repo #1 in this issue)
- `sanitize_for_wire(exc)` helper in roxabi-nats (unblocks Cross-repo #4 in this issue)

Cross-repo #2 (align `roxabi-nats` to `branch=staging`) + Cross-repo #3 (declare `roxabi-contracts` as direct dep) are local — no upstream dependency.

**Comment by MickaelV0** [comment]
## Sub-issues created (2026-05-21)

The 4 PR bundles from the spec slicing are now tracked as sub-issues:

| PR | Sub-issue | Scope | Plan artifact |
|----|-----------|-------|---------------|
| PR-A | #90 | Slices 1+2+3 — cross-repo prep (pin staging, image.SUBJECTS, UnknownEngineError) + CI gates. Closes #86. | `artifacts/plans/89-stage-axis-pr-a-plan.mdx` ✓ |
| PR-B | #91 | Slice 4 — `sanitize_for_wire` on 3 socket-bound paths. Blocked by #90. | TBD (re-run `/plan --issue 91` after PR-A merges) |
| PR-C | #92 | Slice 5 — `TwoPhaseBase` extraction across flux2-klein quant trio. | TBD |
| PR-D | #93 | Slice 6 — `PuLIDMixin` + `_check_nvfp4_requirements`. Closes #89 (final PR). | TBD |

PR-C and PR-D are independent of each other and of PR-B; coordinate at rebase time per spec file-coupling caveat.
